### PR TITLE
obs-ffmpeg: Implement texture encoding for VAAPI

### PR DESCRIPTION
**Pull request by nowrep**
### Description

Implement texture encoding for VAAPI

### Motivation and Context

Based on #8312

### How Has This Been Tested?

Tested with mesa driver.

### Types of changes

- Performance enhancement (non-breaking change which improves efficiency)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by kkartaltepe** [comment]
Encoder crashes/fails on intel when using bframes. Unfortunately ffmpeg encoder dying this way hangs obs... Otherwise it seems to work.

**Comment by nowrep** [comment]
Not sure why would it cause issues with bframes, it is copying into VA surfaces allocated by ffmpeg which should keep them until they are used and only release them after.

Can you post full log?

**Comment by kkartaltepe** [comment]
Sorry I forgot to mention the most important part that it was with AV1, which does not support bframes iirc. Confirmed this happens with the existing  master so its unrelated to this PR.

**Comment by nowrep** [comment]
So that looks like intel driver issue. Both obs and ffmpeg checks if the driver support bframes for given profile.

File: plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c [hunk 1]
```diff
@@ -44,6 +45,9 @@
 #include "vaapi-utils.h"
 #include "obs-ffmpeg-formats.h"
 
+#include <va/va_drmcommon.h>
+#include <libdrm/drm_fourcc.h>
```
Comment (tytan652) [inline plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:49]: Libdrm needs to be added in CMake as a dependency

**Comment by kkartaltepe** [comment]
Got my multi gpu device, and I guess im lucky and I'm not seeing the fallback warning so I presume its working.

**Review by kkartaltepe — APPROVED** [review]

File: plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c (hunk 1 above)
Comment (nowrep) [reply plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:49]: Only linux-pipewire uses libdrm.

File: plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c [hunk 2]
```diff
@@ -619,11 +761,63 @@ static bool vaapi_encode(void *data, struct encoder_frame *frame,
...
+	ret = av_hwframe_transfer_data(hwframe, enc->vframe, 0);
+	if (ret < 0) {
+		warn("vaapi_encode: failed to upload hw frame: %s",
+		     av_err2str(ret));
+		goto fail;
+	}
+
+	ret = av_frame_copy_props(hwframe, enc->vframe);
+	if (ret < 0) {
+		warn("vaapi_encode: failed to copy props to hw frame: %s",
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:813]: The function name is `vaapi_encode_copy`. I'd generally expect the log message prefix to match the function name.

File: plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c [hunk 3]
```diff
@@ -632,6 +826,59 @@ static bool vaapi_encode(void *data, struct encoder_frame *frame,
...
+
+	gs_flush();
+
+	obs_leave_graphics();
+
+	enc->vframe->pts = pts;
+
+	ret = av_frame_copy_props(surface.frame, enc->vframe);
+	if (ret < 0) {
+		warn("vaapi_encode: failed to copy props to hw frame: %s",
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:866]: The function name is `vaapi_encode_tex`. I'd generally expect the log message prefix to match the function name.

**Review by RytoEX — CHANGES_REQUESTED** [review]
I believe this will also need changes to `plugins/obs-ffmpeg/CMakeLists.txt` or this will only build when using legacy CMake, which is due for deprecation and eventual removal.

**Comment by nowrep** [comment]
Added libdrm dep to non-legacy cmake and changed the function names in logging.

**Review by RytoEX — APPROVED** [review]
My nits were addressed. Hopefully someone on Linux can approve/merge.

**Review by Lain-B — APPROVED** [review]

**Pull request merged**